nydus-image: support small chunks mergence

hangvane opened the pull request:
Add the support of merging small file chunks into one batch chunk, for `ConversionType::DirectoryToRafs`, `ConversionType::EStargzToRafs`, `ConversionType::TargzToRafs`, and `ConversionType::TarToRafs`.

# basic usage

Add the `--batch-size` arg to command to enable chunk mergence for supported conversion types:

```shell
nydus-image create --bootstrap ~/bootstrap --blob-dir ~/blobs ~/source --batch-size 0x100000

nydus-image create --type estargz-rafs --bootstrap ~/bootstrap --blob-dir ~/blobs ~/source.tar.gz --batch-size 0x100000

nydus-image create --type targz-rafs --bootstrap ~/bootstrap --blob-dir ~/blobs ~/source.tar.gz --batch-size 0x100000

nydus-image create --type tar-rafs --bootstrap ~/bootstrap --blob-dir ~/blobs ~/source.tar --batch-size 0x100000
```
# benchmarks

- 4K-aligned: append zeros to chunk data buffer so that the data to be compressed is 4K aligned.
- continuous: a big chunk between small chunks will cut the batch chunk.
- 1/2 filtered: do not merge chunks with their uncompressed size bigger than 1/2 batch size.

| repositories | tags | 4K-aligned | continuous | 1/2 filtered | size (MiB) | % | 
|--|--|:--:|:--:|:--:|--:|--:|
| wordpress | latest | - | - | - | 207.36 | - |
| wordpress | latest-rafs | - | - | - | 211.17 | +1.84% |
| wordpress | latest-rafs-batch-64K | ✖ | ✖ | ✖ | 205.63 | -0.83% |
| wordpress | latest-rafs-batch-64K | ✖ | ✖ | ✔ | 205.62 | -0.84% |
| wordpress | latest-rafs-batch-64K | ✖ | ✔ | ✖ | 205.76 | -0.77% |
| wordpress | latest-rafs-batch-64K | ✔ | ✖ | ✖ | 206.17 | -0.57% |
| wordpress | latest-rafs-batch-64K | ✔ | ✖ | ✔ | 206.16 | -0.58% |
| wordpress | latest-rafs-batch-64K | ✖ | ✔ | ✔ | 205.85 | -0.72% |
| wordpress | latest-rafs-batch-64K | ✔ | ✔ | ✖ | 206.28 | -0.52% |
| wordpress | latest-rafs-batch-64K | ✔ | ✔ | ✔ | 206.36 | -0.48% |
| wordpress | latest-rafs-batch-128K | ✖ | ✖ | ✖ | 204.01 | -1.62% |
| wordpress | latest-rafs-batch-128K | ✖ | ✖ | ✔ | 203.91 | -1.66% |
| wordpress | latest-rafs-batch-128K | ✖ | ✔ | ✖ | 204.06 | -1.59% |
| wordpress | latest-rafs-batch-128K | ✔ | ✖ | ✖ | 204.41 | -1.42% |
| wordpress | latest-rafs-batch-128K | ✔ | ✖ | ✔ | 204.41 | -1.42% |
| wordpress | latest-rafs-batch-128K | ✖ | ✔ | ✔ | 204.19 | -1.53% |
| wordpress | latest-rafs-batch-128K | ✔ | ✔ | ✖ | 204.48 | -1.39% |
| wordpress | latest-rafs-batch-128K | ✔ | ✔ | ✔ | 204.61 | -1.32% |
| wordpress | latest-rafs-batch-256K | ✖ | ✖ | ✖ | 201.39 | -2.88% |
| wordpress | latest-rafs-batch-256K | ✖ | ✖ | ✔ | 201.28 | -2.93% |
| wordpress | latest-rafs-batch-256K | ✖ | ✔ | ✖ | 201.44 | -2.85% |
| wordpress | latest-rafs-batch-256K | ✔ | ✖ | ✖ | 201.92 | -2.62% |
| wordpress | latest-rafs-batch-256K | ✔ | ✖ | ✔ | 201.77 | -2.69% |
| wordpress | latest-rafs-batch-256K | ✖ | ✔ | ✔ | 201.45 | -2.85% |
| wordpress | latest-rafs-batch-256K | ✔ | ✔ | ✖ | 201.89 | -2.64% |
| wordpress | latest-rafs-batch-256K | ✔ | ✔ | ✔ | 201.96 | -2.61% |
| wordpress | latest-rafs-batch-512K | ✖ | ✖ | ✖ | 199.23 | -3.92% |
| wordpress | latest-rafs-batch-512K | ✖ | ✖ | ✔ | 199.32 | -3.88% |
| wordpress | latest-rafs-batch-512K | ✖ | ✔ | ✖ | 199.26 | -3.90% |
| wordpress | latest-rafs-batch-512K | ✔ | ✖ | ✖ | 199.35 | -3.86% |
| wordpress | latest-rafs-batch-512K | ✔ | ✖ | ✔ | 199.23 | -3.92% |
| wordpress | latest-rafs-batch-512K | ✖ | ✔ | ✔ | 199.07 | -4.00% |
| wordpress | latest-rafs-batch-512K | ✔ | ✔ | ✖ | 199.43 | -3.82% |
| wordpress | latest-rafs-batch-512K | ✔ | ✔ | ✔ | 199.50 | -3.79% |
| wordpress | latest-rafs-batch-1M | ✖ | ✖ | ✖ | 196.65 | -5.16% |
| wordpress | latest-rafs-batch-1M | ✖ | ✖ | ✔ | 196.67 | -5.16% |
| wordpress | latest-rafs-batch-1M | ✖ | ✔ | ✖ | 196.67 | -5.15% |
| wordpress | latest-rafs-batch-1M | ✔ | ✖ | ✖ | 197.31 | -4.84% |
| wordpress | latest-rafs-batch-1M | ✔ | ✖ | ✔ | 197.25 | -4.87% |
| wordpress | latest-rafs-batch-1M | ✖ | ✔ | ✔ | 196.71 | -5.14% |
| wordpress | latest-rafs-batch-1M | ✔ | ✔ | ✖ | 197.34 | -4.83% |
| wordpress | latest-rafs-batch-1M | ✔ | ✔ | ✔ | 197.41 | -4.80% |
| wordpress | php8.2-fpm-alpine | - | - | - | 98.31 | - |
| wordpress | php8.2-fpm-alpine-rafs | - | - | - | 99.15 | +0.85% |
| wordpress | php8.2-fpm-alpine-rafs-batch-64K | ✖ | ✖ | ✖ | 97.04 | -1.30% |
| wordpress | php8.2-fpm-alpine-rafs-batch-64K | ✖ | ✖ | ✔ | 97.01 | -1.32% |
| wordpress | php8.2-fpm-alpine-rafs-batch-64K | ✖ | ✔ | ✖ | 97.10 | -1.23% |
| wordpress | php8.2-fpm-alpine-rafs-batch-64K | ✔ | ✖ | ✖ | 97.23 | -1.10% |
| wordpress | php8.2-fpm-alpine-rafs-batch-64K | ✔ | ✖ | ✔ | 97.23 | -1.10% |
| wordpress | php8.2-fpm-alpine-rafs-batch-64K | ✖ | ✔ | ✔ | 97.13 | -1.21% |
| wordpress | php8.2-fpm-alpine-rafs-batch-64K | ✔ | ✔ | ✖ | 97.30 | -1.03% |
| wordpress | php8.2-fpm-alpine-rafs-batch-64K | ✔ | ✔ | ✔ | 97.34 | -0.99% |
| wordpress | php8.2-fpm-alpine-rafs-batch-128K | ✖ | ✖ | ✖ | 96.54 | -1.80% |
| wordpress | php8.2-fpm-alpine-rafs-batch-128K | ✖ | ✖ | ✔ | 96.48 | -1.86% |
| wordpress | php8.2-fpm-alpine-rafs-batch-128K | ✖ | ✔ | ✖ | 96.58 | -1.76% |
| wordpress | php8.2-fpm-alpine-rafs-batch-128K | ✔ | ✖ | ✖ | 96.68 | -1.67% |
| wordpress | php8.2-fpm-alpine-rafs-batch-128K | ✔ | ✖ | ✔ | 96.66 | -1.69% |
| wordpress | php8.2-fpm-alpine-rafs-batch-128K | ✖ | ✔ | ✔ | 96.63 | -1.71% |
| wordpress | php8.2-fpm-alpine-rafs-batch-128K | ✔ | ✔ | ✖ | 96.71 | -1.63% |
| wordpress | php8.2-fpm-alpine-rafs-batch-128K | ✔ | ✔ | ✔ | 96.76 | -1.58% |
| wordpress | php8.2-fpm-alpine-rafs-batch-256K | ✖ | ✖ | ✖ | 95.91 | -2.45% |
| wordpress | php8.2-fpm-alpine-rafs-batch-256K | ✖ | ✖ | ✔ | 95.89 | -2.46% |
| wordpress | php8.2-fpm-alpine-rafs-batch-256K | ✖ | ✔ | ✖ | 95.92 | -2.43% |
| wordpress | php8.2-fpm-alpine-rafs-batch-256K | ✔ | ✖ | ✖ | 96.02 | -2.34% |
| wordpress | php8.2-fpm-alpine-rafs-batch-256K | ✔ | ✖ | ✔ | 96.03 | -2.32% |
| wordpress | php8.2-fpm-alpine-rafs-batch-256K | ✖ | ✔ | ✔ | 95.98 | -2.37% |
| wordpress | php8.2-fpm-alpine-rafs-batch-256K | ✔ | ✔ | ✖ | 96.02 | -2.33% |
| wordpress | php8.2-fpm-alpine-rafs-batch-256K | ✔ | ✔ | ✔ | 96.11 | -2.24% |
| wordpress | php8.2-fpm-alpine-rafs-batch-512K | ✖ | ✖ | ✖ | 95.23 | -3.14% |
| wordpress | php8.2-fpm-alpine-rafs-batch-512K | ✖ | ✖ | ✔ | 95.35 | -3.01% |
| wordpress | php8.2-fpm-alpine-rafs-batch-512K | ✖ | ✔ | ✖ | 95.27 | -3.10% |
| wordpress | php8.2-fpm-alpine-rafs-batch-512K | ✔ | ✖ | ✖ | 95.35 | -3.01% |
| wordpress | php8.2-fpm-alpine-rafs-batch-512K | ✔ | ✖ | ✔ | 95.43 | -2.93% |
| wordpress | php8.2-fpm-alpine-rafs-batch-512K | ✖ | ✔ | ✔ | 95.42 | -2.94% |
| wordpress | php8.2-fpm-alpine-rafs-batch-512K | ✔ | ✔ | ✖ | 95.31 | -3.05% |
| wordpress | php8.2-fpm-alpine-rafs-batch-512K | ✔ | ✔ | ✔ | 95.49 | -2.88% |
| wordpress | php8.2-fpm-alpine-rafs-batch-1M | ✖ | ✖ | ✖ | 94.66 | -3.72% |
| wordpress | php8.2-fpm-alpine-rafs-batch-1M | ✖ | ✖ | ✔ | 94.61 | -3.76% |
| wordpress | php8.2-fpm-alpine-rafs-batch-1M | ✖ | ✔ | ✖ | 94.59 | -3.78% |
| wordpress | php8.2-fpm-alpine-rafs-batch-1M | ✔ | ✖ | ✖ | 94.94 | -3.43% |
| wordpress | php8.2-fpm-alpine-rafs-batch-1M | ✔ | ✖ | ✔ | 94.67 | -3.71% |
| wordpress | php8.2-fpm-alpine-rafs-batch-1M | ✖ | ✔ | ✔ | 94.64 | -3.73% |
| wordpress | php8.2-fpm-alpine-rafs-batch-1M | ✔ | ✔ | ✖ | 94.87 | -3.50% |
| wordpress | php8.2-fpm-alpine-rafs-batch-1M | ✔ | ✔ | ✔ | 94.91 | -3.46% |
| python | latest | - | - | - | 335.56 | - |
| python | latest-rafs | - | - | - | 324.83 | - 3.20% |
| python | latest-rafs-batch-64K | ✖ | ✖ | ✖ | 317.84 | -5.28% |
| python | latest-rafs-batch-64K | ✖ | ✖ | ✔ | 317.75 | -5.31% |
| python | latest-rafs-batch-64K | ✖ | ✔ | ✖ | 317.92 | -5.26% |
| python | latest-rafs-batch-64K | ✔ | ✖ | ✖ | 318.48 | -5.09% |
| python | latest-rafs-batch-64K | ✔ | ✖ | ✔ | 318.38 | -5.12% |
| python | latest-rafs-batch-64K | ✖ | ✔ | ✔ | 318.08 | -5.21% |
| python | latest-rafs-batch-64K | ✔ | ✔ | ✖ | 318.59 | -5.06% |
| python | latest-rafs-batch-64K | ✔ | ✔ | ✔ | 318.67 | -5.03% |
| python | latest-rafs-batch-128K | ✖ | ✖ | ✖ | 315.59 | -5.95% |
| python | latest-rafs-batch-128K | ✖ | ✖ | ✔ | 315.40 | -6.01% |
| python | latest-rafs-batch-128K | ✖ | ✔ | ✖ | 315.63 | -5.94% |
| python | latest-rafs-batch-128K | ✔ | ✖ | ✖ | 315.95 | -5.85% |
| python | latest-rafs-batch-128K | ✔ | ✖ | ✔ | 315.94 | -5.85% |
| python | latest-rafs-batch-128K | ✖ | ✔ | ✔ | 315.73 | -5.91% |
| python | latest-rafs-batch-128K | ✔ | ✔ | ✖ | 316.09 | -5.80% |
| python | latest-rafs-batch-128K | ✔ | ✔ | ✔ | 316.23 | -5.76% |
| python | latest-rafs-batch-256K | ✖ | ✖ | ✖ | 312.31 | -6.93% |
| python | latest-rafs-batch-256K | ✖ | ✖ | ✔ | 312.01 | -7.02% |
| python | latest-rafs-batch-256K | ✖ | ✔ | ✖ | 312.42 | -6.90% |
| python | latest-rafs-batch-256K | ✔ | ✖ | ✖ | 312.66 | -6.82% |
| python | latest-rafs-batch-256K | ✔ | ✖ | ✔ | 312.47 | -6.88% |
| python | latest-rafs-batch-256K | ✖ | ✔ | ✔ | 312.54 | -6.86% |
| python | latest-rafs-batch-256K | ✔ | ✔ | ✖ | 312.81 | -6.78% |
| python | latest-rafs-batch-256K | ✔ | ✔ | ✔ | 312.97 | -6.73% |
| python | latest-rafs-batch-512K | ✖ | ✖ | ✖ | 309.06 | -7.90% |
| python | latest-rafs-batch-512K | ✖ | ✖ | ✔ | 308.97 | -7.92% |
| python | latest-rafs-batch-512K | ✖ | ✔ | ✖ | 309.06 | -7.90% |
| python | latest-rafs-batch-512K | ✔ | ✖ | ✖ | 309.23 | -7.85% |
| python | latest-rafs-batch-512K | ✔ | ✖ | ✔ | 309.19 | -7.86% |
| python | latest-rafs-batch-512K | ✖ | ✔ | ✔ | 308.86 | -7.96% |
| python | latest-rafs-batch-512K | ✔ | ✔ | ✖ | 309.38 | -7.80% |
| python | latest-rafs-batch-512K | ✔ | ✔ | ✔ | 309.49 | -7.77% |
| python | latest-rafs-batch-1M | ✖ | ✖ | ✖ | 304.29 | -9.32% |
| python | latest-rafs-batch-1M | ✖ | ✖ | ✔ | 304.67 | -9.21% |
| python | latest-rafs-batch-1M | ✖ | ✔ | ✖ | 304.31 | -9.31% |
| python | latest-rafs-batch-1M | ✔ | ✖ | ✖ | 304.62 | -9.22% |
| python | latest-rafs-batch-1M | ✔ | ✖ | ✔ | 304.90 | -9.14% |
| python | latest-rafs-batch-1M | ✖ | ✔ | ✔ | 304.74 | -9.18% |
| python | latest-rafs-batch-1M | ✔ | ✔ | ✖ | 304.55 | -9.24% |
| python | latest-rafs-batch-1M | ✔ | ✔ | ✔ | 305.08 | -9.08% |
| python | alpine3.17 | - | - | - | 18.64 | - |
| python | alpine3.17-rafs | - | - | - | 20.01 | +7.35% |
| python | alpine3.17-rafs-batch-64K | ✖ | ✖ | ✖ | 19.23 | +3.16% |
| python | alpine3.17-rafs-batch-64K | ✖ | ✖ | ✔ | 19.22 | +3.11% |
| python | alpine3.17-rafs-batch-64K | ✖ | ✔ | ✖ | 19.24 | +3.22% |
| python | alpine3.17-rafs-batch-64K | ✔ | ✖ | ✖ | 19.36 | +3.88% |
| python | alpine3.17-rafs-batch-64K | ✔ | ✖ | ✔ | 19.35 | +3.84% |
| python | alpine3.17-rafs-batch-64K | ✖ | ✔ | ✔ | 19.27 | +3.38% |
| python | alpine3.17-rafs-batch-64K | ✔ | ✔ | ✖ | 19.36 | +3.88% |
| python | alpine3.17-rafs-batch-64K | ✔ | ✔ | ✔ | 19.37 | +3.93% |
| python | alpine3.17-rafs-batch-128K | ✖ | ✖ | ✖ | 19.02 | +2.04% |
| python | alpine3.17-rafs-batch-128K | ✖ | ✖ | ✔ | 19.00 | +1.93% |
| python | alpine3.17-rafs-batch-128K | ✖ | ✔ | ✖ | 19.02 | +2.07% |
| python | alpine3.17-rafs-batch-128K | ✔ | ✖ | ✖ | 19.11 | +2.55% |
| python | alpine3.17-rafs-batch-128K | ✔ | ✖ | ✔ | 19.10 | +2.50% |
| python | alpine3.17-rafs-batch-128K | ✖ | ✔ | ✔ | 19.03 | +2.11% |
| python | alpine3.17-rafs-batch-128K | ✔ | ✔ | ✖ | 19.13 | +2.64% |
| python | alpine3.17-rafs-batch-128K | ✔ | ✔ | ✔ | 19.13 | +2.63% |
| python | alpine3.17-rafs-batch-256K | ✖ | ✖ | ✖ | 18.76 | +0.64% |
| python | alpine3.17-rafs-batch-256K | ✖ | ✖ | ✔ | 18.75 | +0.59% |
| python | alpine3.17-rafs-batch-256K | ✖ | ✔ | ✖ | 18.77 | +0.70% |
| python | alpine3.17-rafs-batch-256K | ✔ | ✖ | ✖ | 18.85 | +1.17% |
| python | alpine3.17-rafs-batch-256K | ✔ | ✖ | ✔ | 18.83 | +1.05% |
| python | alpine3.17-rafs-batch-256K | ✖ | ✔ | ✔ | 18.78 | +0.76% |
| python | alpine3.17-rafs-batch-256K | ✔ | ✔ | ✖ | 18.86 | +1.19% |
| python | alpine3.17-rafs-batch-256K | ✔ | ✔ | ✔ | 18.86 | +1.18% |
| python | alpine3.17-rafs-batch-512K | ✖ | ✖ | ✖ | 18.55 | -0.44% |
| python | alpine3.17-rafs-batch-512K | ✖ | ✖ | ✔ | 18.53 | -0.59% |
| python | alpine3.17-rafs-batch-512K | ✖ | ✔ | ✖ | 18.55 | -0.44% |
| python | alpine3.17-rafs-batch-512K | ✔ | ✖ | ✖ | 18.67 | +0.19% |
| python | alpine3.17-rafs-batch-512K | ✔ | ✖ | ✔ | 18.66 | +0.11% |
| python | alpine3.17-rafs-batch-512K | ✖ | ✔ | ✔ | 18.53 | -0.54% |
| python | alpine3.17-rafs-batch-512K | ✔ | ✔ | ✖ | 18.68 | +0.25% |
| python | alpine3.17-rafs-batch-512K | ✔ | ✔ | ✔ | 18.70 | +0.35% |
| python | alpine3.17-rafs-batch-1M | ✖ | ✖ | ✖ | 18.38 | -1.40% |
| python | alpine3.17-rafs-batch-1M | ✖ | ✖ | ✔ | 18.37 | -1.45% |
| python | alpine3.17-rafs-batch-1M | ✖ | ✔ | ✖ | 18.38 | -1.39% |
| python | alpine3.17-rafs-batch-1M | ✔ | ✖ | ✖ | 18.36 | -1.49% |
| python | alpine3.17-rafs-batch-1M | ✔ | ✖ | ✔ | 18.36 | -1.50% |
| python | alpine3.17-rafs-batch-1M | ✖ | ✔ | ✔ | 18.38 | -1.40% |
| python | alpine3.17-rafs-batch-1M | ✔ | ✔ | ✖ | 18.36 | -1.47% |
| python | alpine3.17-rafs-batch-1M | ✔ | ✔ | ✔ | 18.36 | -1.47% |
| node | latest | - | - | - | 353.79 | - |
| node | latest-rafs | - | - | - | 335.46 | - 5.18% |
| node | latest-rafs-batch-64K | ✖ | ✖ | ✖ | 328.16 | -7.25% |
| node | latest-rafs-batch-64K | ✖ | ✖ | ✔ | 328.07 | -7.27% |
| node | latest-rafs-batch-64K | ✖ | ✔ | ✖ | 328.24 | -7.22% |
| node | latest-rafs-batch-64K | ✔ | ✖ | ✖ | 328.87 | -7.04% |
| node | latest-rafs-batch-64K | ✔ | ✖ | ✔ | 328.79 | -7.07% |
| node | latest-rafs-batch-64K | ✖ | ✔ | ✔ | 328.40 | -7.18% |
| node | latest-rafs-batch-64K | ✔ | ✔ | ✖ | 328.98 | -7.01% |
| node | latest-rafs-batch-64K | ✔ | ✔ | ✔ | 329.07 | -6.99% |
| node | latest-rafs-batch-128K | ✖ | ✖ | ✖ | 325.92 | -7.88% |
| node | latest-rafs-batch-128K | ✖ | ✖ | ✔ | 325.75 | -7.93% |
| node | latest-rafs-batch-128K | ✖ | ✔ | ✖ | 325.98 | -7.86% |
| node | latest-rafs-batch-128K | ✔ | ✖ | ✖ | 326.32 | -7.76% |
| node | latest-rafs-batch-128K | ✔ | ✖ | ✔ | 326.33 | -7.76% |
| node | latest-rafs-batch-128K | ✖ | ✔ | ✔ | 326.07 | -7.84% |
| node | latest-rafs-batch-128K | ✔ | ✔ | ✖ | 326.46 | -7.73% |
| node | latest-rafs-batch-128K | ✔ | ✔ | ✔ | 326.58 | -7.69% |
| node | latest-rafs-batch-256K | ✖ | ✖ | ✖ | 322.75 | -8.78% |
| node | latest-rafs-batch-256K | ✖ | ✖ | ✔ | 322.43 | -8.86% |
| node | latest-rafs-batch-256K | ✖ | ✔ | ✖ | 322.84 | -8.75% |
| node | latest-rafs-batch-256K | ✔ | ✖ | ✖ | 323.10 | -8.68% |
| node | latest-rafs-batch-256K | ✔ | ✖ | ✔ | 322.92 | -8.73% |
| node | latest-rafs-batch-256K | ✖ | ✔ | ✔ | 322.91 | -8.73% |
| node | latest-rafs-batch-256K | ✔ | ✔ | ✖ | 323.25 | -8.63% |
| node | latest-rafs-batch-256K | ✔ | ✔ | ✔ | 323.36 | -8.60% |
| node | latest-rafs-batch-512K | ✖ | ✖ | ✖ | 319.57 | -9.67% |
| node | latest-rafs-batch-512K | ✖ | ✖ | ✔ | 319.57 | -9.67% |
| node | latest-rafs-batch-512K | ✖ | ✔ | ✖ | 319.57 | -9.67% |
| node | latest-rafs-batch-512K | ✔ | ✖ | ✖ | 319.83 | -9.60% |
| node | latest-rafs-batch-512K | ✔ | ✖ | ✔ | 319.82 | -9.60% |
| node | latest-rafs-batch-512K | ✖ | ✔ | ✔ | 319.40 | -9.72% |
| node | latest-rafs-batch-512K | ✔ | ✔ | ✖ | 319.96 | -9.56% |
| node | latest-rafs-batch-512K | ✔ | ✔ | ✔ | 320.07 | -9.53% |
| node | latest-rafs-batch-1M | ✖ | ✖ | ✖ | 315.13 | -10.93% |
| node | latest-rafs-batch-1M | ✖ | ✖ | ✔ | 315.51 | -10.82% |
| node | latest-rafs-batch-1M | ✖ | ✔ | ✖ | 315.19 | -10.91% |
| node | latest-rafs-batch-1M | ✔ | ✖ | ✖ | 315.50 | -10.82% |
| node | latest-rafs-batch-1M | ✔ | ✖ | ✔ | 315.79 | -10.74% |
| node | latest-rafs-batch-1M | ✖ | ✔ | ✔ | 315.61 | -10.79% |
| node | latest-rafs-batch-1M | ✔ | ✔ | ✖ | 315.52 | -10.82% |
| node | latest-rafs-batch-1M | ✔ | ✔ | ✔ | 315.95 | -10.70% |
| node | 19-alpine | - | - | - | 51.47 | - |
| node | 19-alpine-rafs | - | - | - | 43.89 | - 14.73% |
| node | 19-alpine-rafs-batch-64K | ✖ | ✖ | ✖ | 43.10 | -16.27% |
| node | 19-alpine-rafs-batch-64K | ✖ | ✖ | ✔ | 43.09 | -16.28% |
| node | 19-alpine-rafs-batch-64K | ✖ | ✔ | ✖ | 43.10 | -16.25% |
| node | 19-alpine-rafs-batch-64K | ✔ | ✖ | ✖ | 43.22 | -16.03% |
| node | 19-alpine-rafs-batch-64K | ✔ | ✖ | ✔ | 43.22 | -16.03% |
| node | 19-alpine-rafs-batch-64K | ✖ | ✔ | ✔ | 43.12 | -16.23% |
| node | 19-alpine-rafs-batch-64K | ✔ | ✔ | ✖ | 43.21 | -16.04% |
| node | 19-alpine-rafs-batch-64K | ✔ | ✔ | ✔ | 43.23 | -16.01% |
| node | 19-alpine-rafs-batch-128K | ✖ | ✖ | ✖ | 42.94 | -16.57% |
| node | 19-alpine-rafs-batch-128K | ✖ | ✖ | ✔ | 42.93 | -16.59% |
| node | 19-alpine-rafs-batch-128K | ✖ | ✔ | ✖ | 42.94 | -16.57% |
| node | 19-alpine-rafs-batch-128K | ✔ | ✖ | ✖ | 43.03 | -16.40% |
| node | 19-alpine-rafs-batch-128K | ✔ | ✖ | ✔ | 43.02 | -16.42% |
| node | 19-alpine-rafs-batch-128K | ✖ | ✔ | ✔ | 42.94 | -16.57% |
| node | 19-alpine-rafs-batch-128K | ✔ | ✔ | ✖ | 43.03 | -16.40% |
| node | 19-alpine-rafs-batch-128K | ✔ | ✔ | ✔ | 43.02 | -16.40% |
| node | 19-alpine-rafs-batch-256K | ✖ | ✖ | ✖ | 42.78 | -16.89% |
| node | 19-alpine-rafs-batch-256K | ✖ | ✖ | ✔ | 42.77 | -16.90% |
| node | 19-alpine-rafs-batch-256K | ✖ | ✔ | ✖ | 42.77 | -16.90% |
| node | 19-alpine-rafs-batch-256K | ✔ | ✖ | ✖ | 42.86 | -16.72% |
| node | 19-alpine-rafs-batch-256K | ✔ | ✖ | ✔ | 42.86 | -16.73% |
| node | 19-alpine-rafs-batch-256K | ✖ | ✔ | ✔ | 42.78 | -16.89% |
| node | 19-alpine-rafs-batch-256K | ✔ | ✔ | ✖ | 42.88 | -16.68% |
| node | 19-alpine-rafs-batch-256K | ✔ | ✔ | ✔ | 42.88 | -16.68% |
| node | 19-alpine-rafs-batch-512K | ✖ | ✖ | ✖ | 42.67 | -17.09% |
| node | 19-alpine-rafs-batch-512K | ✖ | ✖ | ✔ | 42.67 | -17.10% |
| node | 19-alpine-rafs-batch-512K | ✖ | ✔ | ✖ | 42.67 | -17.09% |
| node | 19-alpine-rafs-batch-512K | ✔ | ✖ | ✖ | 42.76 | -16.92% |
| node | 19-alpine-rafs-batch-512K | ✔ | ✖ | ✔ | 42.76 | -16.92% |
| node | 19-alpine-rafs-batch-512K | ✖ | ✔ | ✔ | 42.67 | -17.09% |
| node | 19-alpine-rafs-batch-512K | ✔ | ✔ | ✖ | 42.74 | -16.95% |
| node | 19-alpine-rafs-batch-512K | ✔ | ✔ | ✔ | 42.74 | -16.95% |
| node | 19-alpine-rafs-batch-1M | ✖ | ✖ | ✖ | 42.59 | -17.25% |
| node | 19-alpine-rafs-batch-1M | ✖ | ✖ | ✔ | 42.58 | -17.27% |
| node | 19-alpine-rafs-batch-1M | ✖ | ✔ | ✖ | 42.61 | -17.21% |
| node | 19-alpine-rafs-batch-1M | ✔ | ✖ | ✖ | 42.69 | -17.05% |
| node | 19-alpine-rafs-batch-1M | ✔ | ✖ | ✔ | 42.69 | -17.06% |
| node | 19-alpine-rafs-batch-1M | ✖ | ✔ | ✔ | 42.61 | -17.21% |
| node | 19-alpine-rafs-batch-1M | ✔ | ✔ | ✖ | 42.66 | -17.11% |
| node | 19-alpine-rafs-batch-1M | ✔ | ✔ | ✔ | 42.66 | -17.11% |

This PR is related to https://github.com/dragonflyoss/image-service/issues/884, https://github.com/dragonflyoss/image-service/issues/885, https://github.com/dragonflyoss/Dragonfly2/issues/1858

hangvane commented:
/retest

adamqqqplay commented:
Thanks for your contribution! Maybe you could squash these commits?

hangvane commented:
> Thanks for your contribution! Maybe you could squash these commits?

@adamqqqplay  Thanks for your advice!  Would you please explain why squashing commits is needed?  I'm a beginner of github, and I'm afraid of messing up this PR by mistake.

adamqqqplay commented:
> > Thanks for your contribution! Maybe you could squash these commits?
> 
> @adamqqqplay Thanks for your advice! Would you please explain why squashing commits is needed? I'm a beginner of github, and I'm afraid of messing up this PR by mistake.

@hangvane In this repo, we use merge to process PRs, and all your commits will go into the master branch. Too many commits in one PR will affect the readability of the git history of the master branch, so we recommend only a small number of commits in each PR.

hsiangkao commented:
> > > Thanks for your contribution! Maybe you could squash these commits?
> > 
> > 
> > @adamqqqplay Thanks for your advice! Would you please explain why squashing commits is needed? I'm a beginner of github, and I'm afraid of messing up this PR by mistake.
> 
> @hangvane In this repo, we use merge to process PRs, and all your commits will go into the master branch. Too many commits in one PR will affect the readability of the git history of the master branch, so we recommend only a small number of commits in each PR.

No, keeping each commit doing one thing is a good behavior, please don’t mislead new contributors here, also see:
https://www.kernel.org/doc/html/latest/process/submitting-patches.html

The main problem here is back merging, we should avoid these merge commits.

hangvane commented:
> > > > Thanks for your contribution! Maybe you could squash these commits?
> > > 
> > > 
> > > @adamqqqplay Thanks for your advice! Would you please explain why squashing commits is needed? I'm a beginner of github, and I'm afraid of messing up this PR by mistake.
> > 
> > 
> > @hangvane In this repo, we use merge to process PRs, and all your commits will go into the master branch. Too many commits in one PR will affect the readability of the git history of the master branch, so we recommend only a small number of commits in each PR.
> 
> No, keeping each commit doing one thing is a good behavior, please don’t mislead new contributors here, also see: https://www.kernel.org/doc/html/latest/process/submitting-patches.html
> 
> The main problem here is back merging, we should avoid these merge commits.

@hsiangkao Thanks! I've removed the back merging commits.

hsiangkao commented:
> > > > > Thanks for your contribution! Maybe you could squash these commits?
> > > > 
> > > > 
> > > > @adamqqqplay Thanks for your advice! Would you please explain why squashing commits is needed? I'm a beginner of github, and I'm afraid of messing up this PR by mistake.
> > > 
> > > 
> > > @hangvane In this repo, we use merge to process PRs, and all your commits will go into the master branch. Too many commits in one PR will affect the readability of the git history of the master branch, so we recommend only a small number of commits in each PR.
> > 
> > 
> > No, keeping each commit doing one thing is a good behavior, please don’t mislead new contributors here, also see: https://www.kernel.org/doc/html/latest/process/submitting-patches.html
> > The main problem here is back merging, we should avoid these merge commits.
> 
> @hsiangkao Yes, what I mean above is similar to yours, which is to write the PR more concisely. It's not that only one commit is allowed, but that the original commit information is too messy.

@hangvane please help squash `[nydus-image: add BATCH flag](https://github.com/dragonflyoss/image-service/pull/1202/commits/765854f3efb99851f031caf2ad894c4523d86fd4)` and `https://github.com/dragonflyoss/image-service/pull/1202/commits/28b4abc57aae8bc24c6a7826be633cab6a245818` if possible, no need to add a new flag and then fix this.

adamqqqplay commented:
> > > > > > Thanks for your contribution! Maybe you could squash these commits?
> > > > > 
> > > > > 
> > > > > @adamqqqplay Thanks for your advice! Would you please explain why squashing commits is needed? I'm a beginner of github, and I'm afraid of messing up this PR by mistake.
> > > > 
> > > > 
> > > > @hangvane In this repo, we use merge to process PRs, and all your commits will go into the master branch. Too many commits in one PR will affect the readability of the git history of the master branch, so we recommend only a small number of commits in each PR.
> > > 
> > > 
> > > No, keeping each commit doing one thing is a good behavior, please don’t mislead new contributors here, also see: https://www.kernel.org/doc/html/latest/process/submitting-patches.html
> > > The main problem here is back merging, we should avoid these merge commits.
> > 
> > 
> > @hsiangkao Yes, what I mean above is similar to yours, which is to write the PR more concisely. It's not that only one commit is allowed, but that the original commit information is too messy.
> 
> @hangvane please help squash `[nydus-image: add BATCH flag](https://github.com/dragonflyoss/image-service/pull/1202/commits/765854f3efb99851f031caf2ad894c4523d86fd4)` and `https://github.com/dragonflyoss/image-service/pull/1202/commits/28b4abc57aae8bc24c6a7826be633cab6a245818` if possible, no need to add a new flag and then fix this.

In addition, this commit `nydus-image: remove debug outputs` also needs squash, because the content it removes is introduced by this PR. It can be kept in your personal repo, but there is no need to commit it upstream.

hangvane commented:
> > > > > > > Thanks for your contribution! Maybe you could squash these commits?
> > > > > > 
> > > > > > 
> > > > > > @adamqqqplay Thanks for your advice! Would you please explain why squashing commits is needed? I'm a beginner of github, and I'm afraid of messing up this PR by mistake.
> > > > > 
> > > > > 
> > > > > @hangvane In this repo, we use merge to process PRs, and all your commits will go into the master branch. Too many commits in one PR will affect the readability of the git history of the master branch, so we recommend only a small number of commits in each PR.
> > > > 
> > > > 
> > > > No, keeping each commit doing one thing is a good behavior, please don’t mislead new contributors here, also see: https://www.kernel.org/doc/html/latest/process/submitting-patches.html
> > > > The main problem here is back merging, we should avoid these merge commits.
> > > 
> > > 
> > > @hsiangkao Yes, what I mean above is similar to yours, which is to write the PR more concisely. It's not that only one commit is allowed, but that the original commit information is too messy.
> > 
> > 
> > @hangvane please help squash `[nydus-image: add BATCH flag](https://github.com/dragonflyoss/image-service/pull/1202/commits/765854f3efb99851f031caf2ad894c4523d86fd4)` and `https://github.com/dragonflyoss/image-service/pull/1202/commits/28b4abc57aae8bc24c6a7826be633cab6a245818` if possible, no need to add a new flag and then fix this.
> 
> In addition, this commit `nydus-image: remove debug outputs` also needs squash, because the content it removes is introduced by this PR. It can be kept in your personal repo, but there is no need to commit it upstream.

@adamqqqplay @hsiangkao I've squashed redundant commits mentioned above. Thank you very much for your patient guidance!

jiangliu commented on src/bin/nydus-image/main.rs:
```diff
@@ -264,6 +264,14 @@ fn prepare_cmd_args(bti_string: &'static str) -> App {
+                .arg(
+                    Arg::new("batch-size")
+                        .long("batch-size")
+                        .help("Set the size of batch data chunks, must be power of two, between 0x1000-0x1000000 or be zero:")
```
Set the batch size to merge small chunks, must be power of two, between 0x1000-0x1000000 or be zero.

hangvane replied:
Changed.

jiangliu commented on src/bin/nydus-image/main.rs:
```diff
@@ -971,6 +982,14 @@ impl Command {
         };
 
+        // Unsupported conversionTypes has been checked and filtered by `get_batch_size()`.
+        if build_ctx.batch_size > 0 {
```
We should only enable small chunk merging for RAFS v6, so need to check fsversion.

hangvane replied:
Changed.
I've add fsversion check here. But I'm not sure if it is better to check fsversion in `get_batch_size()`, or move the ConversionType check from `get_batch_size()` to here. Would you please give me some suggestions?

jiangliu replied:
That's better.

hangvane replied:
Sorry, which one is better?

hangvane replied:
Changed.

hangvane commented on storage/src/meta/chunk_info_v2.rs:
```diff
@@ -16,6 +16,7 @@ const CHUNK_V2_FLAG_MASK: u64 = 0xff << 56;
 const CHUNK_V2_FLAG_COMPRESSED: u64 = 0x1 << 56;
 const CHUNK_V2_FLAG_ZRAN: u64 = 0x2 << 56;
 const CHUNK_V2_FLAG_VALID: u64 = 0x3 << 56;
+const CHUNK_V2_FLAG_BATCH: u64 = 0x4 << 56;
```
Changed.

hangvane commented on rafs/src/builder/core/blob.rs:
```diff
@@ -212,7 +236,7 @@ impl Blob {
             let mut hasher = RafsDigest::hasher(digest::Algorithm::Sha256);
             let ci_data = if ctx.blob_features.contains(BlobFeatures::ZRAN) {
-                zran_buf.as_slice()
+                inflate_buf.as_slice()
```
Changed.

hangvane commented on storage/src/meta/mod.rs:
```diff
@@ -1718,6 +1736,9 @@ pub fn format_blob_features(features: BlobFeatures) -> String {
     if features.contains(BlobFeatures::ZRAN) {
         output += "zran ";
     }
+    if features.contains(BlobFeatures::BATCH) {
```
Changed.

hangvane commented on storage/src/meta/mod.rs:
```diff
@@ -1675,6 +1690,9 @@ pub trait BlobMetaChunkInfo {
     /// Check whether the chunk has associated ZRan context data.
     fn is_zran(&self) -> bool;
 
+    /// Check whether the chunk has associated Batch context data.
```
Changed.

hangvane commented on storage/src/meta/mod.rs:
```diff
@@ -55,6 +55,9 @@ pub mod toc;
 pub use zran::{ZranContextGenerator, ZranInflateContext};
 
+mod batch;
+pub use batch::{BatchInflateContext, BatchInflateContextGenerator};
```
Changed. But I'm still questioning whether this change is right, since `ZranContextGenerator` corresponds to `ZranContext` but `BatchContextGenerator` does not have a corresponding struct `BatchContext` currently.

hangvane commented on src/bin/nydus-image/main.rs:
```diff
@@ -264,6 +264,14 @@ fn prepare_cmd_args(bti_string: &'static str) -> App {
+                        .long("batch-size")
+                        .help("Set the size of batch data chunks, must be power of two, between 0x1000-0x1000000 or be zero:")
+                        .required(false)
+                        .hide(true)
```
Seems that it has been hidden already?

hangvane replied:
Changed.

hangvane commented on rafs/src/metadata/mod.rs:
```diff
@@ -383,6 +385,13 @@ impl RafsSuperConfig {
             )));
         }
 
+        if self.batch_size != meta.batch_size {
```
Changed.
But the minimum check is still kept, that treats `batch_size==0` and `batch_size>0` as incompatible configuration.

jiangliu replied:
I'm not sure why that can be supported. One layer with Batch merging enable, another layer with Batch merging disabled.

hangvane replied:
Removed.

hangvane commented on rafs/src/builder/core/blob.rs:
```diff
@@ -143,23 +163,27 @@ impl Blob {
             header.set_separate_blob(true);
-            header.set_ci_zran(false);
-        } else {
-            header.set_separate_blob(false);
```
Why setting values of `separate_blob` and `ci_zran` to `false` is necessary? I'm confused. Isn't their default value `false`? There're many more flags in `BlobCompressionContextHeader`, e.g., `inlined_fs_meta`. If necessary, why we don't set them to  `false` at the same time?

hangvane commented on storage/src/meta/mod.rs:
```diff
@@ -993,13 +1005,15 @@ impl BlobMetaChunkArray {
     }
 
     /// Add an entry of v2 chunk compression information into the array.
+    #[allow(clippy::too_many_arguments)]
```
Seems it's not easy to import `ChunkWrapper` here. The package dependency is reversed that `nydus_rafs` depends on `nydus_storage`. There's no example that imports `nydus_rafs` in `nydus_storage`. Is it technically possible?

jiangliu replied:
Good point, so keep it as is.

jiangliu commented on src/bin/nydus-image/main.rs:
```diff
@@ -971,6 +981,17 @@ impl Command {
 
+        // Unsupported conversionTypes has been checked and filtered by `get_batch_size()`.
+        if build_ctx.batch_size > 0 {
+            if !version.is_v6() {
```
Better to use `if version.is_v5() {`

hangvane replied:
Changed.

jiangliu commented on storage/src/meta/batch.rs:
```diff
@@ -0,0 +1,144 @@
+        chunk.set_batch(true);
+        chunk.set_batch_index(self.contexts.len() as u32);
+        chunk.set_uncompressed_batch_offset(uncompressed_batch_offset);
+        chunk.set_compressed(true); // TODO: set compressed on demand
```
The TODO can be removed.

jiangliu commented on storage/src/meta/batch.rs:
```diff
@@ -0,0 +1,144 @@
+        Ok(chunk)
+    }
+
+    /// Save the batch chunk information to a file.
```
Please help to update the comment, it's not save data to a file.

hangvane replied:
Updated.

jiangliu commented on storage/src/meta/chunk_info_v2.rs:
```diff
@@ -46,17 +47,41 @@ impl BlobChunkInfoV2Ondisk {
 
     pub(crate) fn set_zran_offset(&mut self, offset: u32) {
+        assert!(self.is_zran());
+        let mut data = u64::from_le(self.data) & !0x0000_0000_ffff_ffff;
```
Could you please help to simplify `!0x0000_0000_ffff_ffff` to `0xffff_ffff_0000_0000`?

hangvane replied:
Changed.

jiangliu commented on rafs/src/builder/core/node.rs:
```diff
@@ -387,32 +391,77 @@ impl Node {
+
+            if batch.chunk_data_buf_len() as u32 + uncompressed_size < ctx.batch_size {
+                // Add into current batch chunk directly.
+                batch.append_chunk_data_buf(chunk_data.to_vec().as_mut());
```
Function  `append_chunk_data_buf()` should be refined to take a `&[u8]` instead of a `Vec[u8]` to avoid unnecessary memory copy.

hangvane replied:
Refined.

jiangliu commented on rafs/src/metadata/mod.rs:
```diff
@@ -383,6 +385,13 @@ impl RafsSuperConfig {
             )));
         }
 
+        if (self.batch_size == 0) != (meta.batch_size == 0) {
```
I still think this check is unnecessary.

hangvane replied:
Removed.

hangvane commented on rafs/src/builder/core/node.rs:
```diff
@@ -387,32 +391,77 @@ impl Node {
+                    uncompressed_size,
+                    blob_ctx.uncompressed_batch_offset,
+                )?);
+                blob_ctx.uncompressed_batch_offset += aligned_chunk_size;
```
Changed.
`uncompressed_batch_offset` now represents the offset in buffer (wihtout 4K aligned), and can be easily calculated by `batch.chunk_data_buf_len()`.
For clarity, `uncompressed_batch_offset` is renamed to `uncompressed_offset_in_batch_buf`.

jiangliu commented on rafs/src/builder/core/node.rs:
```diff
@@ -387,32 +391,75 @@ impl Node {
+                chunk_info = Some(batch.generate_chunk_info(
+                    pre_uncompressed_offset,
+                    uncompressed_size,
+                    uncompressed_offset_in_batch_buf,
```
It would be better to omit the parameter `uncompressed_offset_in_batch_buf` from batch.chunk_data_buf if we switch the order to generate chunk info and append data to the buffer.

hangvane replied:
Good idea. Changed.

adam3q reviewed (changes requested):
Please excuse my repetition, commits still need to be tidied up before the code is finally merged to keep the history readable.
My suggestion is to split your PR in as independent modules as possible, not in the order of commits, please refer to: https://softwareengineering.stackexchange.com/questions/263164/why-squash-git-commits-for-pull-requests

hangvane commented:
> Please excuse my repetition, commits still need to be tidied up before the code is finally merged to keep the history readable. My suggestion is to split your PR in as independent modules as possible, not in the order of commits, please refer to: https://softwareengineering.stackexchange.com/questions/263164/why-squash-git-commits-for-pull-requests

@adamqqqplay I'm still confused about why we need PR author to squash all the commits to one before merged. It's reasonable to squash some commits for a clear view to reviewers, but I still think keeping the commits separate is better, especially after code review.

And the reason that too many commits will pollute the history (of the master branch?) seems unreasonable. Depends on the choice of the merger, the commits of PR can be automatically squashed into one merge commit to the master branch, as shown in this figure.

![image](https://user-images.githubusercontent.com/22473909/231942000-dad35494-7107-4845-a68d-99f96c2baa56.png)
https://www.cnblogs.com/xgqfrms/p/16464515.html

​Here are some more examples showing that many merged PR did not squash their commits. This would not pollute the commit history of the master branch.

https://github.com/dragonflyoss/image-service/pull/1171, https://github.com/dragonflyoss/image-service/pull/1176, https://github.com/dragonflyoss/Dragonfly2/pull/2190

hangvane commented:
> Please excuse my repetition, commits still need to be tidied up before the code is finally merged to keep the history readable. My suggestion is to split your PR in as independent modules as possible, not in the order of commits, please refer to: https://softwareengineering.stackexchange.com/questions/263164/why-squash-git-commits-for-pull-requests

@adamqqqplay I've realized that the merge policy of this project is different with d7y. I'll squash all the commits before merged. Thank you!

adamqqqplay commented:
> > Please excuse my repetition, commits still need to be tidied up before the code is finally merged to keep the history readable. My suggestion is to split your PR in as independent modules as possible, not in the order of commits, please refer to: https://softwareengineering.stackexchange.com/questions/263164/why-squash-git-commits-for-pull-requests
> 
> @adamqqqplay I've realized that the merge policy of this project is different with d7y. I'll squash all the commits before merged. Thank you!

First of all thanks for your contribution! I'm sorry I didn't make it clear before, each repo has its own different management methods, in this repo we didn't use the squash and merge method used in your image.

I don't mean that you can't submit multiple commits, but that the log of a partial commit is only meaningful in that PR. When this PR is merged into master, the commit log will be confusing. For example: `nydus-image: fix code review suggestions` and so on.

From a software engineering point of view, a good commit on the master branch is one that works independently and can be efficiently tested, rolled back, or cherry-picked. For me personally, it may be more appropriate for your final PR to contain 2-4 commits. For example, the first few commits fix or refactor existing code, and the last few commits introduce your new feature.

hsiangkao commented:
> > > Please excuse my repetition, commits still need to be tidied up before the code is finally merged to keep the history readable. My suggestion is to split your PR in as independent modules as possible, not in the order of commits, please refer to: https://softwareengineering.stackexchange.com/questions/263164/why-squash-git-commits-for-pull-requests
> > 
> > 
> > @adamqqqplay I've realized that the merge policy of this project is different with d7y. I'll squash all the commits before merged. Thank you!
> 
> First of all thanks for your contribution! I'm sorry I didn't make it clear before, each repo has its own different management methods, in this repo we didn't use the squash and merge method used in your image.
> 
> I don't mean that you can't submit multiple commits, but that the log of a partial commit is only meaningful in that PR. When this PR is merged into master, the commit log will be confusing. For example: `nydus-image: fix code review suggestions` and so on.
> 
> From a software engineering point of view, a good commit on the master branch is one that works independently and can be efficiently tested, rolled back, or cherry-picked. For me personally, it may be more appropriate for your final PR to contain 2-4 commits. For example, the first few commits fix or refactor existing code, and the last few commits introduce your new feature.

Personally, I'd suggest don't too picky to new contributors. Even the current Nydus commits cannot meet your requirements.
I'd be better to sort out these commits in a proper way, but that is not a MUST.

hangvane committed:
storage: check `zran` flag before set `zran` values
Check `zran` flag before set `zran` values.
Refine comments.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
rafs: reuse chunk data compress and write procedure
Refactor `Node::dump_file_chunk()` to reuse data compress and write procedure.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
storage: introduce BatchInflateContext to support batch chunk
Enhance chunk info to support batch chunk.
Introduce BatchInflateContext and generator.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
rafs: enhance builder to support batch chunk
Add `--batch-size` subcommand on nydus-image.
Add build time support of batch chunk.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane commented:
> > > Please excuse my repetition, commits still need to be tidied up before the code is finally merged to keep the history readable. My suggestion is to split your PR in as independent modules as possible, not in the order of commits, please refer to: https://softwareengineering.stackexchange.com/questions/263164/why-squash-git-commits-for-pull-requests
> > 
> > 
> > @adamqqqplay I've realized that the merge policy of this project is different with d7y. I'll squash all the commits before merged. Thank you!
> 
> First of all thanks for your contribution! I'm sorry I didn't make it clear before, each repo has its own different management methods, in this repo we didn't use the squash and merge method used in your image.
> 
> I don't mean that you can't submit multiple commits, but that the log of a partial commit is only meaningful in that PR. When this PR is merged into master, the commit log will be confusing. For example: `nydus-image: fix code review suggestions` and so on.
> 
> From a software engineering point of view, a good commit on the master branch is one that works independently and can be efficiently tested, rolled back, or cherry-picked. For me personally, it may be more appropriate for your final PR to contain 2-4 commits. For example, the first few commits fix or refactor existing code, and the last few commits introduce your new feature.

@adamqqqplay Thanks for your suggestions! I've reorganized the commits. Each commit have passed github workflows in https://github.com/hangvane/image-service/tree/build-batch

I would be grateful if you could have a check.

adam3q reviewed (approved):
Everything looks good, thanks for your great work!

hangvane commented:
@jiangliu A problem about runtime decompressing is considering. Please do not merge this PROBLEM at this time. I'll add more benchmark cases for discussion.

jiangliu commented:
According to the test result, how about the default configuration of:
- 4k aligned = true: simplify runtime handling of batched chunks
- continuous = false: it causes compression ratio decrease without any benefits. 
- 1/2 filter = true: it achieves better compression ratios under most cases

hangvane commented:
> According to the test result, how about the default configuration of:
> 
> * 4k aligned = true: simplify runtime handling of batched chunks
> * continuous = false: it causes compression ratio decrease without any benefits.
> * 1/2 filter = true: it achieves better compression ratios under most cases

@jiangliu The basic idea is that, `continuous = true` is used combined with `4k aligned = true` to simplify runtime handling of batched chunks, that the raw decompressed batch chunk data can be dumped to uncompressed blob data as a whole without the consideration of each small chunk and its chunk info. Since the basic idea is wrong, that we need all the chunk info of small chunks anyway, it seems inappropriate to adopt `continuous = true` and `4k aligned = true`. Unless otherwise motivated, I think it's better not to add these two features and keep the code in this PR as it is.

As for `1/2 filter`, we can add a configuration to give the user the ability to set this feature. But I think just enable this feature hard-coded, keep the code in this PR as what it is currently, is good enough already.

jiangliu commented:
> > According to the test result, how about the default configuration of:
> > 
> > * 4k aligned = true: simplify runtime handling of batched chunks
> > * continuous = false: it causes compression ratio decrease without any benefits.
> > * 1/2 filter = true: it achieves better compression ratios under most cases
> 
> @jiangliu The basic idea is that, `continuous = true` is used combined with `4k aligned = true` to simplify runtime handling of batched chunks, that the raw decompressed batch chunk data can be dumped to uncompressed blob data as a whole without the consideration of each small chunk and its chunk info. Since the basic idea is wrong, that we need all the chunk info of small chunks anyway, it seems inappropriate to adopt `continuous = true` and `4k aligned = true`. Unless otherwise motivated, I think it's better not to add these two features and keep the code in this PR as it is.
> 
> As for `1/2 filter`, we can add a configuration to give the user the ability to set this feature. But I think just enable this feature hard-coded, keep the code in this PR as what it is currently, is good enough already.

It would be better to avoid introducing configuration for `1/2 filter`, the benefit is trivial and keep it simple for users.
